### ISSUE TITLE
read_only mode

### ISSUE THREAD

**Issue by johnnyshields** [gmailgem/gmail#55]
_From @craigulliott on January 22, 2012 22:48_

Firstly, thanks!  I love this gem.

Secondly, here is one observation for my specific use case:

I have a script which is part of a custom CRM, it loops through all messages in a gmail account, and matches them up to a specific contact in our database.

When each message is downloaded from Gmail, it is automatically marked as read.  This is very confusing for the owners of those mailboxes.  There seems to be no method to get the read? or unread? status of a mail message which has just been accessed, otherwise I would simply mark it as unread! after I was finished.  (Appears this is a limitation of the IMAP protocol).

The solution to this is deep inside the documentation for Net:IMAP, and it is to access the mailbox with "examine".   

Here is some code I am using in an initializer which perfectly fits my use case, I am putting it here partly to help anyone who may stumble upon it from google, and partly to suggest the addition of a :read_only => true feature when connecting.

If there is enough demand for this, I'm happy to submit a patch.

Thanks!

```ruby
module Gmail
  module Client
    class Base

      def switch_to_mailbox(mailbox)
        if mailbox
          mailbox = Net::IMAP.encode_utf7(mailbox)
          conn.examine(mailbox)
        end
        @current_mailbox = mailbox
      end

    end
  end
end
```

_Copied from original issue: nu7hatch/gmail#55_

**Comment by johnnyshields** [comment]
_From @jgrevich on January 22, 2012 22:58_

Hi Craig,

I noticed the same problem.  I've had to switch gears to a different project at the moment, but I plan to address this issue by copying the peek functionality found in this library's parent project: ruby-gmail.  Take a look:

```ruby
    # Parsed MIME message object
    def message
      require 'mail'
      request,part = 'RFC822','RFC822'
      request,part = 'BODY.PEEK[]','BODY[]' if @gmail.peek
      _body = @gmail.in_mailbox(@mailbox) { @gmail.imap.uid_fetch(uid, request)[0].attr[part] }
      @message ||= Mail.new(_body)
    end
```

```ruby
# Get messages without marking them as read on the server.
gmail.peek = true
gmail.inbox.emails(:unread, :before => Date.parse("2010-04-20"), :from => "myboss@gmail.com")
```

I haven't had a chance to test it, but it looks like they solved your problem.  We'd appreciate your help if you can submit a patch with similar functionality.


justin

**Comment by johnnyshields** [comment]
_From @codyrobbins on December 7, 2012 0:14_

It looks like this functionality was already implemented by @waseems in pull request #27 (along with some other useful additions), but it wasn't pulled in for some reason.

**Comment by johnnyshields** [comment]
_From @webcracy on October 4, 2013 18:13_

With the message's "envelope", you'll get the information you need without reading the message and therefore achieve what you'll need.
